src/components/pageData.js

```
@@ -0,0 +1,162 @@
+  getDataForPage (res, fileName, fileParams) {
+    return new Promise((resolve, reject) => {
+      let data = ''
+      res.on('data', d => {
+        data += d
+      })
+      res.on('end', () => {
+        try {
+          data = JSON.parse(data)
+          if (!data) return reject(new Error('Nothing returned'))
+          if (data.message) return reject(data.message)
+          const response = this.extractData(data, fileName, fileParams)
+          return resolve(response)
+        } catch (e) {
+          return reject(e)
+        }
+      })
+    })
+  }
+
+  // Call the API per markdown file and get data for each one returned.
+  callAPI (fileName, fileParams) {
+    return new Promise((resolve, reject) => {
+      const request = this.prepareRequest(fileParams)
+      const requestMethod = request.port === '443' ? https : http
+      requestMethod.get(request, res => {
+        this.getDataForPage(res, fileName, fileParams).then(fileData => {
+          // Remove the markdown file from metalsmith as its not an actual page
+          delete this.params.files[ fileName ]
+          if (fileData.response) return resolve(fileData.response)
+          return reject(new Error('No response found'))
+        }).catch(reject)
+      })
+    })
+  }
+
+  makeExtraAPICalls (data, fileParams, callBack) {
+    // Now check for additional requests per page returned from API call
+    // This can be prodlib data based on an SEO object
+    if (!fileParams.dataSource.extras) return callBack()
+    const loop = fileParams.dataSource.extras
+    return Object.keys(loop).map(opt => {
+      if (!loop[ opt ].query) return
+      loop[ opt ].extraFiles = true
+      const option = loop[ opt ].query.match(/<%([^%].*)%>/)
+      const extraPageData = data.map(currentFile => {
+        return new Promise((resolve, reject) => {
+          if (!currentFile.data.pageData[option[1]]) return reject(new Error('No extra options found'))
```

Reject should throw the error in the callback, probably just cleaner to pass the message in perhaps

---

That doesn't conform to standardjs linting if you dont say its an Error